harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -619,32 +1003,66 @@ def _compute_validation_metrics(self) -> workload.Metrics:
-                    "on_validation_step_end is now deprecated, please use on_validation_end instead"
+                    "on_validation_step_end is now deprecated, please use on_validation_end instead."
                 )
                 callback.on_validation_step_end(metrics)
 
         for callback in self.callbacks.values():
             callback.on_validation_end(metrics)
 
-        if not self.is_chief:
-            return {}
+        self.state.last_val = self.state.batches_trained
 
-        # Skip reporting timings if evaluate_full_dataset() was defined.  This is far less common
-        # than evaluate_batch() and we can't know how the user processed their validation data.
-        if self._evaluate_batch_defined():
-            step_duration = time.time() - step_start_time
-            logging.info(
-                det.util.make_timing_log("validated", step_duration, num_inputs, num_batches)
-            )
-        self.metric_writer.on_validation_step_end(self.steps_completed, metrics)
-        return {"num_inputs": num_inputs, "validation_metrics": metrics}
+        if self.is_chief:
+            # Skip reporting timings if evaluate_full_dataset() was defined.  This is far less
+            # common than evaluate_batch() and we can't know how the user processed their
+            # validation data.
+            if self._evaluate_batch_defined():
+                step_duration = time.time() - step_start_time
+                logging.info(
+                    det.util.make_timing_log("validated", step_duration, num_inputs, num_batches)
+                )
+            self.metric_writer.on_validation_step_end(self.state.batches_trained, metrics)
+
+        should_checkpoint = False
+
+        if self.is_chief:
+            if self.local_training:
+                searcher_length = self.max_length
+            else:
+                searcher_length = TrainUnit._from_searcher_unit(
+                    searcher_op.length, self.searcher_unit, self.global_batch_size
+                )
+            if self.searcher_metric_name:
+                searcher_metric = self._check_searcher_metric(metrics)
+            else:
+                searcher_metric = None
+
+            if self._steps_until_complete(searcher_length) < 1 and not searcher_op._completed:
+                searcher_op.report_completed(searcher_metric)
+
+            if self.ckpt_policy == "best" and not self._checkpoint_is_current():
+                best_validation_before = self.core_context.train.get_experiment_best_validation()
+
+            self.core_context.train.report_validation_metrics(self.state.batches_trained, metrics)
+
+            if not self._checkpoint_is_current():
+                if self.ckpt_policy == "all" or (
+                    self.ckpt_policy == "best"
+                    and self._is_best_validation(now=searcher_metric, before=best_validation_before)
+                ):
+                    should_checkpoint = True
```

Comment:
This code is overly complex still.  I think there's a bug where `searcher_op` might be None but you calculate a searcher_length anyway, then you might call `searcher_op.report_completed()` on a None object.

I would recommend you separate these into logical blocks:
- one for validation metrics (actually the previous `if` statement is already all about metrics, so just put report_validation_metrics there)
- one for searcher op handling, which needs to be gated on searcher_op being non-None
- one for checkpoint policy handling

Maybe like this:

```python
# Metric stuff.
if self.is_chief:
    # Skip reporting timings if evaluate_full_dataset() was defined.  This is far less
    # common than evaluate_batch() and we can't know how the user processed their
    # validation data.
    if self._evaluate_batch_defined():
        step_duration = time.time() - step_start_time
        logging.info(
            det.util.make_timing_log("validated", step_duration, num_inputs, num_batches)
        )
    self.metric_writer.on_validation_step_end(self.state.batches_trained, metrics)

    #### I moved this here, since this block is all about metrics
    self.core_context.train.report_validation_metrics(self.state.batches_trained, metrics)

# Searcher operation stuff, only for real searchers.
if self.is_chief and self.searcher_metric_name and searcher_op:
    searcher_length = TrainUnit._from_searcher_unit(
        searcher_op.length, self.searcher_unit, self.global_batch_size
    )
    searcher_metric = self._check_searcher_metric(metrics)
    if self._steps_until_complete(searcher_length) < 1 and not searcher_op._completed:
        searcher_op.report_completed(searcher_metric)

# Checkpoint policy stuff.
should_checkpoint = False
if self.is_chief:
    if self.ckpt_policy == "best" and not self._checkpoint_is_current():
        best_validation_before = self.core_context.train.get_experiment_best_validation()

    if not self._checkpoint_is_current():
        if self.ckpt_policy == "all" or (
            self.ckpt_policy == "best"
            and self._is_best_validation(now=searcher_metric, before=best_validation_before)
        ):
            should_checkpoint = True

should_checkpoint = self.context.distributed.broadcast(should_checkpoint)
if should_checkpoint:
    self._checkpoint(already_exiting=False)
return metrics
```

What you currently have jumps from metric stuff, to checkpoint stuff, to searcher stuff, to metric stuff, then back to checkpoint stuff.

Reply:
did this, mostly. the "only for real searchers" stuff isn't quite only for real searchers. dummy searcher ops have no searcher_metric_name but still should report_completed.